actions/allocate.php

```diff
@@ -75,8 +76,9 @@
 $sorthow = optional_param('sorthow', '', PARAM_ALPHA);
 $options = compact('sortby', 'sorthow', 'perpage', 'page');
 
-// $_POST['allocatables'] comes as array of arrays which is not supported by optional_param_array, however we clean this later in process_data() function
-$formdataarray = isset($_POST['allocatables']) ? $_POST['allocatables'] : [];
+// Variable $_POST['allocatables'] comes as array of arrays which is not supported by optional_param_array.
+// However, we clean this later in process_data() function.
+$dirtyformdata = isset($_POST['allocatables']) ? $_POST['allocatables'] : [];
 
```

Comment:
can we refactor it to not use array of arrays?

Reply:
Absolutely if we think it's worth the extra effort.  We could refactor the form so that it returns a single array with each former nested array having multiple entries in the single array.  Would you rather do that?

Reply:
I assume it might risk breaking stuff as a result? but probably worth doing

Reply:
It would be a little fiddly and not sure it would have a high priority after I added the extra cleaning, also given that page is limited to users with the capability 'mod/coursework:allocate'.   When the form is saved, with the extra cleaning, each nested array is cleaned, with its keys validated and its values cleaned with the core clean_param_array()